commands/migratelxc.go

```diff
@@ -346,3 +370,211 @@ func renameLXDContainers(
 	}
 	return group.Wait()
 }
+
+// startLXDContainers starts the LXD containers that have
+// just been migrated from LXC, or were already migrated
+// but not yet started.
+func startLXDContainers(
+	lxcByHost map[*state.Machine][]*state.Machine,
+	lxdByHost map[*state.Machine]map[string]*lxdContainer,
+	containerNames map[*state.Machine]containerNames,
+) error {
+	running := make(map[string]bool) // keyed by LXD container name
+	for _, containers := range lxdByHost {
+		for name, container := range containers {
+			if container.IsActive() {
+				running[name] = true
+			}
+		}
+	}
+	var group errgroup.Group
+	for host, containers := range lxcByHost {
+		// By this stage, all of the LXC containers
+		// have been migrated to LXD and renamed.
+		// If they don't feature in "lxdByHost",
+		// then they were migrated by this process,
+		// and they're known to not be running.
+		toStart := make([]string, 0, len(containers))
+		for _, container := range containers {
+			newName := containerNames[container].newName
+			if !running[newName] {
+				toStart = append(toStart, newName)
+			}
+		}
+		if len(toStart) == 0 {
+			logger.Debugf("no LXD containers to start on %q", host.Id())
+			continue
+		}
+		logger.Debugf("starting LXD containers on %q: %q", host.Id(), toStart)
+		host := host // copy for closure
+		group.Go(func() error {
+			return errors.Annotatef(
+				StartLXDContainers(toStart, host),
+				"starting LXD containers on %q: %q",
+				host.Id(), toStart,
+			)
+		})
+	}
+	return group.Wait()
+}
+
+// waitLXDContainersReady waits for the LXD containers to have
+// addresses, as recorded in the state database, and be ready to
+// accept SSH connections.
+func waitLXDContainersReady(
+	lxcByHost map[*state.Machine][]*state.Machine,
+	containerNames map[*state.Machine]containerNames,
+	timeout time.Duration,
+) error {
+	containerAddrs := make(map[string]string) // keyed by LXD container name
+	for _, containers := range lxcByHost {
+		for _, container := range containers {
+			addr, err := getMachineAddress(container)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			name := containerNames[container].newName
+			containerAddrs[name] = addr
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	group, ctx := errgroup.WithContext(ctx)
+	const interval = time.Second // time to wait between checks
+
+	waitHostLXDContainersReady := func(host *state.Machine) error {
```

Comment:
I feel like two nested levels of closures and parallel execution are too many. Would it be difficult to extract waitHostLXDContainersReady to a separate method/function so that it can be understood in isolation? Then I can get my head around the fanout that you do there (from a host to its containers), so I can ignore it when understanding the fanout to the hosts.

Actually, does the parallelism need to be two levels? Can't you just flatten them down to a list of containers (wherever they're hosted) and then parallelize the sshing to them? I think that would be easier to follow.